### DISCUSSION TITLE
How to Check LLM's Performance

### DISCUSSION THREAD

**Discussion by DevanshuBrahmbhatt** [microsoft/promptflow#362] [Q&A] [answered]
I am looking for ways to check or evaluate the LLM's performance for various prompts or various inputs. What is the metric for it? If we are chunking the dataset and generating Embeddings then how the chunking strategies will affect the LLM's performance If it is then how to measure it?

**Comment by ttthree** [comment, marked as answer]
The answer is "it depends on your scenario" ... and you will need to pick metric + define dataset(s) to run and calculate metric.
E.g. if you're using the LLM for [classification](https://github.com/microsoft/promptflow/tree/main/examples/flows/standard/web-classification) or [named entity recognition](https://github.com/microsoft/promptflow/tree/main/examples/flows/standard/named-entity-recognition), you can use simple/classical metrics like [accuracy](https://github.com/microsoft/promptflow/tree/main/examples/flows/evaluation/eval-classification-accuracy), [% match](https://github.com/microsoft/promptflow/tree/main/examples/flows/evaluation/eval-entity-match-rate). These usually have a ground-truth to compare with.

If you're using the LLM for more complex scenarios like creating a chatbot then a new set of metric should be define, things like:
- Whether the chatbot is answering questions based on facts and provided context. (people may call it groundedness)
- How intelligent/smart the answer look like, e.g. is it relevant to the question, is it creative and able to impress the user
- How efficient the bot can fulfill the ask from the user
- etc.

Unlike the traditional metrics which can be calculated based on ground-truth, you usually don't have a ground-truth for these metrics. A classical approach will be relying on human review/labeling to decide the performance. Fortunately, models like GPT-4 is good at labelling meaning that if you can define the metrics with clean instructions (which is required anyway even you do human labelling), GPT-4 can do a good job. Take a look at these examples: [groundedness](https://github.com/microsoft/promptflow/blob/main/examples/flows/evaluation/eval-groundedness/gpt_groundedness.md), [perceived intelligence](https://github.com/microsoft/promptflow/blob/main/examples/flows/evaluation/eval-perceived-intelligence/gpt_perceived_intelligence.md) and how they're being used in one of our tutorials - [a chatbot answering questions based on a pdf](https://github.com/microsoft/promptflow/blob/main/examples/tutorials/e2e-development/chat-with-pdf.md)

Note: these evaluation flows and prompts are just for demo purposes, you will have to craft prompts for evaluation of your own app/flow. That's where you might invest a lot on.

Also once you figured these out, prompt flow will be quite helpful for you to update your dev workflow to include the evaluations into CI/CD pipeline etc.

Prompt flow can also help you on testing multiple prompts easily, check out [the variants feature](https://microsoft.github.io/promptflow/concepts/concept-variants.html) and how it's being used in the [web-classification](https://github.com/microsoft/promptflow/tree/main/examples/flows/standard/web-classification) example:
- You can add variant with combination of different prompts and/or different model parameters
- You can run these variants in vs code extension in one shot and look at results
- You can run batch on these variants using SDK/CLI and evaluate them to compare performance
![CleanShot 2023-09-11 at 05 40 38@2x](https://github.com/microsoft/promptflow/assets/18907/215c0689-308b-4077-92c8-ddd5ff9c737b)
<img width="1643" alt="CleanShot 2023-09-11 at 05 55 21@2x" src="https://github.com/microsoft/promptflow/assets/18907/dda5ddf2-75c4-4bc2-9526-0dc5a776a305">

**Comment by DevanshuBrahmbhatt** [comment]
Thank you so much for your response. I can share my scenario. I am creating a Q&A Chatbot, (Whether the chatbot is answering questions based on facts and providing context. (People may call it groundedness) ) I have the normal approach of Taking 

`PDF-> Raw Text-> Chunks-> Embeddings-> Vector Similarity-> Create Context ->  Prompt -> LLM's Result`

However, I'm interested in creating a testing workflow to assess how each of these stages can impact the response from the LLM. Additionally, I'd like to measure their impact using specific performance parameters

**Reply by PossumXI** [reply]
I want to create a AI assistant using machine learning and NLP to generate predictive holistic medicine consumption based on several factors that are automatically tracked to several open API's. Any suggestions on how to go about this?

**Reply by anoexpected** [reply]
To generate personalized holistic medicine recommendations using machine learning and NLP, I would suggest the following approach:

1. **Collect data** on factors that can influence holistic medicine needs, such as demographic info, vital signs, lab tests, genetics, diet, symptoms, etc. You mentioned tracking these automatically through open APIs, which is a great start. Look for APIs from wearables, electronic health records, health tracking apps, etc.

2. **Build a knowledge base** of holistic medicines and their intended uses by scraping websites, published research, practitioner forums, etc. Use NLP techniques like named entity recognition to extract medicine names, conditions they treat, ingredients, etc.

3. **Develop a model** that takes a patient's current data profile as input and outputs ranked holistic medicine recommendations. This could be a rule-based system, Bayesian model, neural network, or other ML/NLP algorithm. 

4. **Train and evaluate the model** on real patient data. Apply techniques like k-fold cross-validation to avoid overfitting. Assess predictive accuracy, recall, precision, etc.

5. **Implement a conversational interface** so patients can interact with the model through an assistant like you. Use NLP to parse patient queries and extract key details to provide personalized, context-aware recommendations.

6. **Monitor real-world performance** and continue training the model on new data over time. Explore ensembling multiple models to improve robustness.

**Reply by boorge** [reply]
Building an AI assistant for predictive holistic medicine consumption involves a multi-faceted approach. Begin by selecting robust machine learning frameworks like TensorFlow or PyTorch. Utilize NLP models such as BERT or GPT-3 for understanding user input. Integrate your system with open APIs that provide relevant health data.

**Data Collection:** Gather diverse datasets including health records, lifestyle choices, and holistic medicine effectiveness.

**Preprocessing:** Clean and format the data, ensuring compatibility for machine learning algorithms.

**Model Development:** Train your ML models on historical data, allowing the system to learn patterns and correlations.

**API Integration:** Connect your AI assistant to open health APIs for real-time information on holistic remedies and medical trends.

**User Interface:** Design a user-friendly interface for seamless interaction, allowing users to input data and receive personalized predictions.

**Remember**, ongoing refinement is crucial for accuracy. Regularly update your models and maintain API connections to ensure your AI assistant provides reliable predictions.

**Comment by Charlotte-br560** [comment]
To evaluate LLM performance across prompts or inputs, consider using metrics like perplexity, BLEU score, or F1 score depending on your task. Chunking strategies can impact performance; assess this by comparing results across different chunking approaches. For measuring it, analyze metrics on validation or test sets.

**Comment by kinthaiofficial** [comment]
Checking LLM performance in PromptFlow is best approached at multiple levels:

**1. Output quality metrics (LLM-as-judge)**
PromptFlow has built-in evaluation flows for common quality dimensions: groundedness, relevance, coherence, fluency. Run these against a sample of your production outputs:

```python
# In a eval flow
from promptflow.core import tool

@tool
def evaluate_groundedness(answer: str, context: str) -> float:
    # Use an LLM to score 1-5 whether answer is grounded in context
    ...
```

**2. Latency and throughput metrics**
PromptFlow traces log per-node execution time. Export these to Azure Monitor or your APM of choice. Key metrics: P50/P95/P99 latency per LLM node, tokens/second, cache hit rate if using semantic caching.

**3. Cost tracking**
PromptFlow's trace data includes token usage per run. Aggregate this in Azure Monitor with:
```
customMetrics 
| where name == "promptflow.tokens_consumed"
| summarize sum(value) by bin(timestamp, 1h)
```

**4. Regression detection**
For ongoing quality monitoring: run your eval flows on a fixed "golden dataset" after each prompt change. Alert if quality scores drop more than X% from baseline.

**5. Batch evaluation vs online evaluation**
Batch: run eval flows on historical data, get aggregate quality scores.
Online: sample a % of live requests, run them through eval flow asynchronously, aggregate in dashboards.

What's the primary performance concern — quality degradation, latency, or cost?
